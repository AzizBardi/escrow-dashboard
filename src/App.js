import { ThemeProvider } from "@mui/material/styles";

import theme from "./theme";
import Main from "./components/Main";
import "./App.css";
import { ApolloProvider } from "@apollo/client";
import { getClient } from "./queries";
import { networkMap } from "./constants";
import AppContext from "../src/AppContext";
import { useState } from "react";

function App() {
  const [network, setNetwork] = useState("polygon");

  return (
    <AppContext.Provider value={{ network, setNetwork }}>
      <ApolloProvider client={getClient(networkMap[network].graphqlClientUrl)}>
        <ThemeProvider theme={theme}>
          <Main />
        </ThemeProvider>
      </ApolloProvider>
    </AppContext.Provider>
  );
}

export default App;
