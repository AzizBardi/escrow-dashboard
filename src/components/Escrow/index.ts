export { EscrowContainer as Escrow } from './Escrow';
