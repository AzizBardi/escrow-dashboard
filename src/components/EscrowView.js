import { Divider, Typography } from "@mui/material";
import Card from "@mui/material/Card";
import CardContent from "@mui/material/CardContent";

export default function Escrow(props) {
  const {
    status,
    address,
    repOracleAddr,
    repOracleStake,
    recOracleAddr,
    recOracleStake,
    balance,
    manifestUrl,
    finalResultsUrl,
  } = props;

  return (
    <Card variant="outlined">
      <CardContent>
          <CardTextBlock title="Address" value={address} />
          <CardTextBlock title="Status" value={status} />
          <CardTextBlock title="Reputation Oracle" value={repOracleAddr} />
          <CardTextBlock title="Reputation Oracle stake(%)" value={repOracleStake} />
          <CardTextBlock title="Recording Oracle" value={recOracleAddr} />
          <CardTextBlock title="Recording Oracle stake(%)" value={recOracleStake} />
          <CardTextBlock title="Balance(HMT)" value={balance} />
          <CardTextBlock title="Manifest" value={manifestUrl} />
          <CardTextBlock title="Final Results" value={finalResultsUrl} />
      </CardContent>
    </Card>
  );
}

function CardTextBlock({ title, value }) {
  return (
    <>
      <Divider textAlign="center" sx={{mt: 1}}>
        <Typography variant="body2" color="text.secondary">
          {title}
        </Typography>
      </Divider>
      <Typography variant="body2" textAlign="center">
        {value}
      </Typography>
    </>
  );
}
