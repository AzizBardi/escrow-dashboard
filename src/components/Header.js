import { useRecoilState, useSetRecoilState } from "recoil";
import AppBar from "@mui/material/AppBar";
import Box from "@mui/material/Box";
import Button from "@mui/material/Button";
import Typography from "@mui/material/Typography";
import Toolbar from "@mui/material/Toolbar";

import {ReactComponent as MetamaskLogo} from '../metamask.svg';
import { walletState } from "../store/wallet";
import { isPolygon } from "../constants";
import { enableWeb3, switchToPolygon, getWeb3 } from "../WalletProvider";

function Header() {
  const [wallet] = useRecoilState(walletState);
  const setWallet = useSetRecoilState(walletState);

  const onConnect = async () => {
    const web3 = getWeb3();
    await enableWeb3();

    window.ethereum.on('chainChanged', (chainChanged) => {
      setWallet(oldWallet => ({ ...oldWallet, networkChainId: chainChanged }));
    });

    window.ethereum.on('accountsChanged', (accounts) => {
      if (accounts.length > 0) {
        setWallet(oldWallet => ({ ...oldWallet, address: accounts[0]}));
      } else {
        setWallet({ networkChainId: '', address: '', isConnected: false})
      }
    });

    const networkChainId = await window.ethereum.request({
      method: "eth_chainId",
    });
    const account = (await web3.eth.getAccounts())[0];
    setWallet((oldWallet) => ({
      ...oldWallet,
      address: account,
      networkChainId,
      isConnected: true,
    }));
  };

  return (
    <Box sx={{ flexGrow: 1 }}>
      <AppBar position="static">
        <Toolbar>
          <Typography
            variant="h6"
            component="div"
            color="white"
            sx={{
              flexGrow: 1,
              paddingLeft: 2,
              paddingBottom: 2,
              paddingTop: 2,
            }}
          >
            Escrow scanner
          </Typography>
          <Button
            variant="contained"
            startIcon={<MetamaskLogo />}
            color="info"
            disabled={wallet.isConnected}
            onClick={onConnect}
          >
            Connect
          </Button>
          <Button
            variant="contained"
            color="info"
            disabled={!wallet.isConnected || isPolygon(wallet.networkChainId)}
            onClick={switchToPolygon}
            sx={{ marginLeft: "4px" }}
          >
            Switch To Polygon
          </Button>
        </Toolbar>
      </AppBar>
    </Box>
  );
}

export default Header;
