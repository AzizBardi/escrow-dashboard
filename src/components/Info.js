import { useRecoilState } from "recoil";
import Typography from "@mui/material/Typography";
import Card from "@mui/material/Card";
import CardContent from "@mui/material/CardContent";
import Divider from "@mui/material/Divider";

import { walletState } from "../store/wallet";
import { escrowAddressState } from "../store/escrow";
import { chainIdToTitle, chainIdToScannerUrl } from "../constants";

export default function Info() {
  const [wallet] = useRecoilState(walletState);
  const [escrowAddr] = useRecoilState(escrowAddressState);
  const networkScannerUrl = wallet.networkChainId && chainIdToScannerUrl[wallet.networkChainId];
  const escrowUrl = `${networkScannerUrl}/address/${escrowAddr}`;

  return (
    <Card variant="outlined" sx={{ width: "450px" }}>
      <CardContent>
        <Divider textAlign="center">
          <Typography color="text.secondary" variant="body2">
            Network
          </Typography>
        </Divider>
        <Typography variant="body2" textAlign="center">
          {chainIdToTitle[wallet.networkChainId]}
        </Typography>
        <Divider textAlign="center" sx={{ mt: 1 }}>
          <Typography color="text.secondary" variant="body2">
            Escrow
          </Typography>
        </Divider>
        <Typography variant="body2" textAlign="center">
          {escrowAddr && <a href={escrowUrl} rel="noopener noreferrer" target="_blank"> {escrowAddr}</a>}
        </Typography>
      </CardContent>
    </Card>
  );
}
