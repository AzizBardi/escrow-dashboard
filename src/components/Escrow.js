import {useEffect, useState} from 'react';
import { useRecoilState, useRecoilValue } from 'recoil';

import {escrowAddressState} from '../store/escrow';
import {isConnectedSelector} from '../store/wallet';
import { getWeb3 } from '../WalletProvider';
import EscrowView from './EscrowView';
import EscrowABI from '../contracts/EscrowAbi.json';

const statusesMap = ['Launched', 'Pending', 'Partial', 'Paid', 'Complete', 'Cancelled'];

export default function EscrowContainer() {
    const [escrowAddr] = useRecoilState(escrowAddressState);
    const isConnected = useRecoilValue(isConnectedSelector);
    const [escrowStatus, setEscrowStatus] = useState('');

    const [reputationOracle, setReputationOracle] = useState('');
    const [reputationOracleStake, setReputationOracleStake] = useState('');
  
    const [recordingOracle, setRecordingOracle] = useState('');
    const [recordingOracleStake, setRecordingOracleStake] = useState('');
  
    const [manifestUrl, setManifestUrl] = useState('');
    const [finalResultsUrl, setFinalResultsUrl] = useState('');
    const [balance, setBalance] = useState('');
  
  

    useEffect(() => {
        if (!isConnected || !escrowAddr) {
            return null;
        }
        async function setupEscrow() {
            const web3 = getWeb3();
            const Escrow = new web3.eth.Contract(EscrowABI, escrowAddr);
            const escrowSt = await Escrow.methods.status().call();
            setEscrowStatus(statusesMap[escrowSt]);
        
            const recOracleAddr = await Escrow.methods.recordingOracle().call();
            setRecordingOracle(recOracleAddr);
        
            const recOracleStake = await Escrow.methods.recordingOracleStake().call();
            setRecordingOracleStake(recOracleStake);
        
            const repOracleAddr = await Escrow.methods.reputationOracle().call();
            setReputationOracle(repOracleAddr);
        
            const repOracleStake = await Escrow.methods.reputationOracleStake().call();
            setReputationOracleStake(repOracleStake);
        
            const finalResults = await Escrow.methods.finalResultsUrl().call();
            setFinalResultsUrl(finalResults);
        
            const manifest = await Escrow.methods.manifestUrl().call();
            setManifestUrl(manifest);
        
            const balance = await Escrow.methods.getBalance().call();
            setBalance(web3.utils.fromWei(balance, 'ether'));
        
        }

        setupEscrow();

    }, [escrowAddr, isConnected]);

    return (
        <EscrowView
          status={escrowStatus}
          address={escrowAddr}
          repOracleAddr={reputationOracle}
          recOracleAddr={recordingOracle}
          repOracleStake={reputationOracleStake}
          recOracleStake={recordingOracleStake}
          balance={balance}
          manifestUrl={manifestUrl}
          finalResultsUrl={finalResultsUrl}

        />
    )
}
