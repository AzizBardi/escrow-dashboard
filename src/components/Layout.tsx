import { Box } from '@mui/material';
import * as React from 'react';
import Header from './Header';

interface ILayout {
  children: React.ReactNode;
}

const Layout: React.FC<ILayout> = ({ children }): React.ReactElement => (
  <Box
    sx={{
      marginTop: '120px',
    }}
  >
    <Header />
    {children}
  </Box>
);

export default Layout;
