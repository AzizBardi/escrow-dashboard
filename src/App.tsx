import * as React from 'react';

import { ThemeProvider } from '@mui/material/styles';
import { ApolloProvider } from '@apollo/client';
import { BrowserRouter as Router, Routes, Route } from 'react-router-dom';

import theme from './theme';
import './App.css';
import { getClient } from './queries';
import { networkMap } from './constants';
import AppNetworkContext from './AppNetworkContext';

import { routes as appRoutes } from './routes';
import Layout from './components/Layout';

const App: React.FC = (): React.ReactElement => {
  const [network, setNetwork] = React.useState<string>('polygon');

  return (
    <AppNetworkContext.Provider value={{ network, setNetwork }}>
      <ApolloProvider client={getClient(networkMap[network].graphqlClientUrl)}>
        <ThemeProvider theme={theme}>
          <Router>
            <Layout>
              <Routes>
                {appRoutes.map((route) => (
                  <Route
                    key={route.key}
                    path={route.path}
                    element={<route.component />}
                  />
                ))}
              </Routes>
            </Layout>
          </Router>
        </ThemeProvider>
      </ApolloProvider>
    </AppNetworkContext.Provider>
  );
};

export default App;
