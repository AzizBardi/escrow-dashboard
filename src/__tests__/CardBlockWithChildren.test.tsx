/* eslint-disable jsx-a11y/aria-role */
import * as React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { CardBlockWithChildren } from 'src/components';

describe('when rendered CardBlockWithChildren component', () => {
  it('should render passed `children` prop', () => {
    const { getByTestId } = render(
      <CardBlockWithChildren>
        <div data-testid="root">
          <div data-testid="parent">
            <div data-testid="child">content</div>
          </div>
        </div>
      </CardBlockWithChildren>
    );

    const root = getByTestId('root');
    const parent = getByTestId('parent');
    const child = getByTestId('child');
    expect(root).toContainElement(parent);
    expect(parent).toContainElement(child);
    expect(child).not.toContainElement(parent);
  });
});

describe('when rendered CardBlockWithChildren component with a `text` prop', () => {
  it('should render passed `children` prop with passed `Test` text prop', () => {
    const { getByText } = render(
      <CardBlockWithChildren>
        <div data-testid="root">
          <span>Test</span>
        </div>
      </CardBlockWithChildren>
    );

    expect(getByText(/Test/)).toBeInTheDocument();
  });
});

it('CardBlockWithChildren component renders correctly, corresponds to the snapshot', () => {
  const tree = renderer
    .create(
      <CardBlockWithChildren>
        <div data-testid="root">
          <span>Test</span>
        </div>
      </CardBlockWithChildren>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
