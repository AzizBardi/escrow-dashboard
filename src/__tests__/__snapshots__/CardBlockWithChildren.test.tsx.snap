// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardBlockWithChildren component renders correctly, corresponds to the snapshot 1`] = `
<p
  className="MuiTypography-root MuiTypography-body2 css-1ypfif-MuiTypography-root"
>
  <div
    data-testid="root"
  >
    <span>
      Test
    </span>
  </div>
</p>
`;
